File: frontend/lib/merkle.ts
/**
 * TypeScript port of `staccana_genesis::merkle` — produces byte-identical hashes
 * to the Rust reference impl at `genesis/src/merkle.rs`.
 *
 * Hash function: SHA-256 (via WebCrypto). Domain separation byte 0x00 for
 * leaves, 0x01 for internal nodes — prevents second-preimage attacks.
 *
 * Leaves are sorted by pubkey ascending (raw 32-byte lex order — which is what
 * `solana_program::pubkey::Pubkey::cmp` does). Odd layers duplicate the last
 * hash. The single remaining hash is the root.
 *
 * Verified byte-equal to the Rust impl via fixtures in `tests/merkle.test.ts`.
 */

import { PublicKey } from "@solana/web3.js";

import { LEAF_DOMAIN, NODE_DOMAIN } from "./staccana";

/** A claimable account from the snapshot. Mirrors `ClaimableLeaf` in Rust. */
export interface ClaimableLeaf {
  pubkey: PublicKey;
  /** Lamports as a bigint (u64 in Rust, encoded little-endian). */
  lamports: bigint;
}

/** A Merkle inclusion proof for one leaf. Mirrors `InclusionProof` in Rust. */
export interface InclusionProof {
  pubkey: PublicKey;
  lamports: bigint;
  /** Sibling hashes from leaf-level upward. Each is exactly 32 bytes. */
  proof: Uint8Array[];
  /**
   * Packed bit flags. Length is `ceil(proof.len() / 8)` bytes.
   * Bit i (LSB-first within each byte, byte 0 first) is 1 iff the sibling at
   * level i is on the right (running hash on the left).
   */
  proofFlags: Uint8Array;
  /** The Merkle root reconstructed from this proof. */
  root: Uint8Array;
}

/** Sync SHA-256 implementation that does not depend on WebCrypto's async API. */
async function sha256(input: Uint8Array): Promise<Uint8Array> {
  const digest = await crypto.subtle.digest("SHA-256", input);
  return new Uint8Array(digest);
}

/** Concatenate Uint8Array chunks into a single buffer. */
function concat(chunks: Uint8Array[]): Uint8Array {
  let total = 0;
  for (const c of chunks) total += c.length;
  const out = new Uint8Array(total);
  let off = 0;
  for (const c of chunks) {
    out.set(c, off);
    off += c.length;
  }
  return out;
}

/** Encode a u64 as 8 little-endian bytes. */
export function u64LeBytes(n: bigint): Uint8Array {
  if (n < 0n || n > 0xff_ff_ff_ff_ff_ff_ff_ffn) {
    throw new RangeError(`u64 out of range: ${n}`);
  }
  const out = new Uint8Array(8);
  let v = n;
  for (let i = 0; i < 8; i++) {
    out[i] = Number(v & 0xffn);
    v >>= 8n;
  }
  return out;
}

/**
 * Hash a single claimable leaf: `SHA256(0x00 || pubkey || lamports.to_le_bytes())`.
 * Matches `ClaimableLeaf::hash` in Rust.
 */
export async function leafHash(leaf: ClaimableLeaf): Promise<Uint8Array> {
  return sha256(concat([new Uint8Array([LEAF_DOMAIN]), leaf.pubkey.toBytes(), u64LeBytes(leaf.lamports)]));
}

/**
 * Hash an internal node: `SHA256(0x01 || left || right)`. Matches `node_hash`
 * in Rust.
 */
export async function nodeHash(left: Uint8Array, right: Uint8Array): Promise<Uint8Array> {
  return sha256(concat([new Uint8Array([NODE_DOMAIN]), left, right]));
}

/** Lex-compare two pubkeys. Matches `Pubkey::cmp` in Rust (raw byte order). */
function pubkeyCmp(a: PublicKey, b: PublicKey): number {
  const x = a.toBytes();
  const y = b.toBytes();
  for (let i = 0; i < 32; i++) {
    if (x[i] < y[i]) return -1;
    if (x[i] > y[i]) return 1;
  }
  return 0;
}

/** Sort leaves ascending by pubkey. Returns a new sorted copy. */
export function sortLeaves(leaves: ClaimableLeaf[]): ClaimableLeaf[] {
  return [...leaves].sort((a, b) => pubkeyCmp(a.pubkey, b.pubkey));
}

/** A 32-byte zero hash; the Rust impl returns this for an empty input. */
const ZERO_HASH = new Uint8Array(32);

/**
 * Build the Merkle root over a set of claimable leaves.
 *
 * Algorithm:
 * 1. Sort leaves ascending by pubkey.
 * 2. Compute leaf hashes.
 * 3. Iteratively reduce: pair adjacent hashes via nodeHash(left, right).
 *    If a layer has odd length, the last hash pairs with itself.
 * 4. The single remaining hash is the root.
 *
 * Empty input => returns 32 bytes of zeros (matches Rust's `Hash::default()`).
 */
export async function buildMerkleRoot(leaves: ClaimableLeaf[]): Promise<Uint8Array> {
  if (leaves.length === 0) return ZERO_HASH;
  const sorted = sortLeaves(leaves);
  let layer = await Promise.all(sorted.map(leafHash));
  while (layer.length > 1) {
    const next: Uint8Array[] = [];
    for (let i = 0; i < layer.length; i += 2) {
      const left = layer[i];
      const right = i + 1 < layer.length ? layer[i + 1] : layer[i];
      next.push(await nodeHash(left, right));
    }
    layer = next;
  }
  return layer[0];
}

/**
 * Build the inclusion proof for `target` against the claimable partition `leaves`.
 *
 * Mirrors `build_inclusion_proof` in `tools/claim-cli/src/proof.rs`. Returns
 * `null` if the target is not in the leaf set.
 *
 * `proofFlags` is the packed bitmap from SPEC §4.1: bit `i` (LSB-first within
 * each byte, byte 0 first) controls level `i`. `0` => sibling on the left
 * (running hash on the right). `1` => sibling on the right (running hash on
 * the left).
 */
export async function buildInclusionProof(
  leaves: ClaimableLeaf[],
  target: PublicKey,
): Promise<InclusionProof | null> {
  if (leaves.length === 0) return null;

  const sorted = sortLeaves(leaves);
  const targetIndex = sorted.findIndex((leaf) => pubkeyCmp(leaf.pubkey, target) === 0);
  if (targetIndex < 0) return null;
  const targetLeaf = sorted[targetIndex];

  let layer = await Promise.all(sorted.map(leafHash));
  let idx = targetIndex;
  const proof: Uint8Array[] = [];
  const siblingOnRightFlags: boolean[] = [];

  while (layer.length > 1) {
    let siblingIdx: number;
    let siblingOnRight: boolean;
    if (idx % 2 === 0) {
      // Even index: sibling is the next slot, or self in the odd-leaf case.
      if (idx + 1 < layer.length) {
        siblingIdx = idx + 1;
        siblingOnRight = true;
      } else {
        siblingIdx = idx;
        siblingOnRight = true;
      }
    } else {
      // Odd index: sibling is the previous slot (on the left).
      siblingIdx = idx - 1;
      siblingOnRight = false;
    }
    proof.push(layer[siblingIdx]);
    siblingOnRightFlags.push(siblingOnRight);

    // Promote layer.
    const next: Uint8Array[] = [];
    for (let i = 0; i < layer.length; i += 2) {
      const left = layer[i];
      const right = i + 1 < layer.length ? layer[i + 1] : layer[i];
      next.push(await nodeHash(left, right));
    }
    idx = Math.floor(idx / 2);
    layer = next;
  }

  return {
    pubkey: targetLeaf.pubkey,
    lamports: targetLeaf.lamports,
    proof,
    proofFlags: packBits(siblingOnRightFlags),
    root: layer[0],
  };
}

/**
 * Recompute the root from a proof. Useful as a self-check before submitting
 * a transaction. Matches `InclusionProof::recomputed_root` in Rust.
 */
export async function recomputeRoot(proof: InclusionProof): Promise<Uint8Array> {
  let running = await sha256(
    concat([new Uint8Array([LEAF_DOMAIN]), proof.pubkey.toBytes(), u64LeBytes(proof.lamports)]),
  );
  for (let i = 0; i < proof.proof.length; i++) {
    const sibling = proof.proof[i];
    const siblingOnRight = bitIsSet(proof.proofFlags, i);
    if (siblingOnRight) {
      running = await nodeHash(running, sibling);
    } else {
      running = await nodeHash(sibling, running);
    }
  }
  return running;
}

/** Pack a list of bools into a little-endian-bit byte vector. */
export function packBits(bits: boolean[]): Uint8Array {
  const nBytes = Math.ceil(bits.length / 8);
  const out = new Uint8Array(nBytes);
  for (let i = 0; i < bits.length; i++) {
    if (bits[i]) {
      out[Math.floor(i / 8)] |= 1 << (i % 8);
    }
  }
  return out;
}

/** Read bit i from a packed bit vector. */
export function bitIsSet(bytes: Uint8Array, i: number): boolean {
  return ((bytes[Math.floor(i / 8)] >> (i % 8)) & 1) === 1;
}

/** Hex-encode a Uint8Array. Helpful for debugging fixture comparisons. */
export function toHex(bytes: Uint8Array): string {
  return Array.from(bytes, (b) => b.toString(16).padStart(2, "0")).join("");
}

/** Hex-decode a string into a Uint8Array. Throws on odd length / non-hex chars. */
export function fromHex(hex: string): Uint8Array {
  const clean = hex.startsWith("0x") ? hex.slice(2) : hex;
  if (clean.length % 2 !== 0) {
    throw new Error(`fromHex: odd-length input (${clean.length})`);
  }
  const out = new Uint8Array(clean.length / 2);
  for (let i = 0; i < out.length; i++) {
    const byte = parseInt(clean.slice(i * 2, i * 2 + 2), 16);
    if (Number.isNaN(byte)) {
      throw new Error(`fromHex: non-hex byte at index ${i}`);
    }
    out[i] = byte;
  }
  return out;
}

/**
 * Derive `proofFlags` from a leaf index. At level i, the sibling is on the
 * right iff the current index is even — so flag bit i = `(idx >> i) & 1 === 0`.
 *
 * Mirrors the index-walking logic inside `buildInclusionProof`, but works
 * without access to the leaf set: we only need the leaf index + the proof
 * length (= number of levels above the leaf).
 *
 * Use this when an edge fn returns the proof siblings + leafIndex but omits
 * the packed bitmap (the bitmap is purely a function of leafIndex).
 */
export function deriveProofFlagsFromLeafIndex(
  leafIndex: number,
  proofLen: number,
): Uint8Array {
  const bits: boolean[] = [];
  let idx = leafIndex;
  for (let i = 0; i < proofLen; i++) {
    bits.push(idx % 2 === 0);
    idx = Math.floor(idx / 2);
  }
  return packBits(bits);
}

File: programs/lazy-claim/src/merkle.rs
//! On-chain Merkle proof verification against the embedded `claimable_root`.
//!
//! This is the inclusion-check counterpart to `staccana_genesis::merkle::MerkleTree::build`:
//! given a `(pubkey, lamports)` leaf and a sibling-hash path, recompute the root and
//! compare against the embedded value.
//!
//! ## Domain bytes — MUST stay in sync with `staccana_genesis::merkle`
//!
//! Both crates use SHA-256 (via `solana_program::hash::hashv`) with explicit domain bytes
//! to prevent leaf/internal-node second-preimage attacks:
//!
//! * `0x00` prefix → leaf hash
//! * `0x01` prefix → internal node hash
//!
//! Constants are duplicated here (rather than re-exported from staccana-genesis) so the
//! lazy-claim crate has no dependency on the genesis-builder crate. If the genesis side
//! ever changes either constant, this file MUST change to match — there is a unit test in
//! the integration tier that asserts root equality across both crates.
//!
//! ## Proof flag encoding
//!
//! `proof_flags` is a packed bitfield over `proof.len()` bits, LSB-first within each byte:
//!
//! * bit i = 0 → sibling is on the LEFT of the running hash; combine as `node(sibling, running)`
//! * bit i = 1 → sibling is on the RIGHT of the running hash; combine as `node(running, sibling)`
//!
//! Trailing bits in the final byte beyond `proof.len()` are ignored.

use solana_program::hash::{hashv, Hash};

const LEAF_DOMAIN: u8 = 0x00;
const NODE_DOMAIN: u8 = 0x01;

/// Compute the leaf hash for a `(pubkey, lamports)` claim. Identical to
/// `staccana_genesis::merkle::ClaimableLeaf::hash`.
pub fn leaf_hash(pubkey: &[u8; 32], lamports: u64) -> Hash {
    hashv(&[&[LEAF_DOMAIN], pubkey.as_ref(), &lamports.to_le_bytes()])
}

/// Combine two child hashes into a parent node hash. Identical to the internal
/// `node_hash` used in `staccana_genesis::merkle::MerkleTree::build`.
pub fn node_hash(left: &Hash, right: &Hash) -> Hash {
    hashv(&[&[NODE_DOMAIN], left.as_ref(), right.as_ref()])
}

/// Walk a Merkle proof from `leaf` up the tree. Returns the computed root.
///
/// `proof_flags` packs one bit per sibling, LSB-first. Bit `i` controls level `i`:
/// * `0` ⇒ sibling on left, running hash on right
/// * `1` ⇒ sibling on right, running hash on left
pub fn compute_root(leaf: Hash, proof: &[Hash], proof_flags: &[u8]) -> Hash {
    let mut running = leaf;
    for (i, sibling) in proof.iter().enumerate() {
        let byte = proof_flags[i / 8];
        let bit = (byte >> (i % 8)) & 1;
        running = if bit == 0 {
            node_hash(sibling, &running)
        } else {
            node_hash(&running, sibling)
        };
    }
    running
}

/// Verify a Merkle inclusion proof against `expected_root`.
///
/// Returns `true` iff the recomputed root equals `expected_root` byte-for-byte.
/// Caller is responsible for ensuring `proof_flags.len() >= ceil(proof.len() / 8)`.
pub fn verify_inclusion(
    leaf: Hash,
    proof: &[Hash],
    proof_flags: &[u8],
    expected_root: &Hash,
) -> bool {
    let computed = compute_root(leaf, proof, proof_flags);
    computed.as_ref() == expected_root.as_ref()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn pk(byte: u8) -> [u8; 32] {
        [byte; 32]
    }

    /// Build a tiny tree manually so each test case knows exactly what root to expect.
    /// Mirrors the algorithm in `staccana_genesis::merkle::MerkleTree::build`.
    fn build_root(leaves: &[([u8; 32], u64)]) -> Hash {
        if leaves.is_empty() {
            return Hash::default();
        }
        let mut sorted: Vec<([u8; 32], u64)> = leaves.to_vec();
        sorted.sort_by(|a, b| a.0.cmp(&b.0));
        let mut layer: Vec<Hash> = sorted.iter().map(|(p, l)| leaf_hash(p, *l)).collect();
        while layer.len() > 1 {
            let mut next = Vec::with_capacity(layer.len() / 2 + 1);
            for chunk in layer.chunks(2) {
                let parent = if chunk.len() == 2 {
                    node_hash(&chunk[0], &chunk[1])
                } else {
                    node_hash(&chunk[0], &chunk[0])
                };
                next.push(parent);
            }
            layer = next;
        }
        layer[0]
    }

    /// Construct a proof + flag set for `target_idx` against the canonical
    /// `MerkleTree::build` algorithm. Returns `(proof, packed_flags)`.
    fn build_proof(leaves: &[([u8; 32], u64)], target_idx: usize) -> (Vec<Hash>, Vec<u8>) {
        let mut sorted: Vec<([u8; 32], u64)> = leaves.to_vec();
        sorted.sort_by(|a, b| a.0.cmp(&b.0));
        let mut layer: Vec<Hash> = sorted.iter().map(|(p, l)| leaf_hash(p, *l)).collect();

        let mut proof: Vec<Hash> = Vec::new();
        let mut flag_bits: Vec<u8> = Vec::new(); // 0 or 1 per level
        let mut idx = target_idx;

        while layer.len() > 1 {
            let pair_idx = idx ^ 1;
            let (sibling, sibling_on_right) = if pair_idx < layer.len() {
                // Standard pair.
                (layer[pair_idx], idx % 2 == 0)
            } else {
                // Odd-leaf duplication: our node pairs with itself.
                (layer[idx], true)
            };
            proof.push(sibling);
            flag_bits.push(if sibling_on_right { 1 } else { 0 });

            let mut next = Vec::with_capacity(layer.len() / 2 + 1);
            for chunk in layer.chunks(2) {
                let parent = if chunk.len() == 2 {
                    node_hash(&chunk[0], &chunk[1])
                } else {
                    node_hash(&chunk[0], &chunk[0])
                };
                next.push(parent);
            }
            idx /= 2;
            layer = next;
        }

        // Pack flags LSB-first.
        let mut packed = vec![0u8; (flag_bits.len() + 7) / 8];
        for (i, b) in flag_bits.iter().enumerate() {
            packed[i / 8] |= b << (i % 8);
        }
        (proof, packed)
    }

    #[test]
    fn verify_succeeds_for_valid_proof() {
        let leaves = vec![
            (pk(1), 100u64),
            (pk(2), 200u64),
            (pk(3), 300u64),
            (pk(4), 400u64),
        ];
        let root = build_root(&leaves);

        for (i, (p, l)) in leaves.iter().enumerate() {
            let (proof, flags) = build_proof(&leaves, sorted_index_of(&leaves, p));
            let leaf = leaf_hash(p, *l);
            assert!(
                verify_inclusion(leaf, &proof, &flags, &root),
                "leaf {} (lamports {}) failed verification",
                i,
                l
            );
        }
    }

    #[test]
    fn verify_fails_for_wrong_root() {
        let leaves = vec![(pk(1), 100u64), (pk(2), 200u64)];
        let (proof, flags) = build_proof(&leaves, 0);
        let leaf = leaf_hash(&pk(1), 100);
        let bogus_root = Hash::new_from_array([0xAB; 32]);
        assert!(!verify_inclusion(leaf, &proof, &flags, &bogus_root));
    }

    #[test]
    fn verify_fails_for_tampered_proof() {
        let leaves = vec![
            (pk(1), 100u64),
            (pk(2), 200u64),
            (pk(3), 300u64),
            (pk(4), 400u64),
        ];
        let root = build_root(&leaves);
        let (mut proof, flags) = build_proof(&leaves, 0);

        // Flip a byte in a sibling hash.
        let mut bytes = proof[0].to_bytes();
        bytes[0] ^= 0xFF;
        proof[0] = Hash::new_from_array(bytes);

        let leaf = leaf_hash(&pk(1), 100);
        assert!(!verify_inclusion(leaf, &proof, &flags, &root));
    }

    #[test]
    fn verify_fails_for_wrong_leaf_content() {
        let leaves = vec![(pk(1), 100u64), (pk(2), 200u64)];
        let root = build_root(&leaves);
        let (proof, flags) = build_proof(&leaves, 0);

        // Right pubkey, wrong lamports.
        let bad_leaf = leaf_hash(&pk(1), 999);
        assert!(!verify_inclusion(bad_leaf, &proof, &flags, &root));

        // Wrong pubkey, right lamports.
        let bad_leaf = leaf_hash(&pk(99), 100);
        assert!(!verify_inclusion(bad_leaf, &proof, &flags, &root));
    }

    #[test]
    fn verify_handles_odd_layer_with_self_pairing() {
        // 3 leaves → odd at the bottom. Verify proof for the lone third leaf, which pairs
        // with itself one level up.
        let leaves = vec![(pk(1), 100u64), (pk(2), 200u64), (pk(3), 300u64)];
        let root = build_root(&leaves);
        let (proof, flags) = build_proof(&leaves, 2);
        let leaf = leaf_hash(&pk(3), 300);
        assert!(verify_inclusion(leaf, &proof, &flags, &root));
    }

    #[test]
    fn verify_single_leaf_tree_has_empty_proof() {
        let leaves = vec![(pk(7), 42u64)];
        let root = build_root(&leaves);
        let leaf = leaf_hash(&pk(7), 42);
        assert!(verify_inclusion(leaf, &[], &[], &root));
        // Wrong lamports must fail even with empty proof.
        let bad_leaf = leaf_hash(&pk(7), 43);
        assert!(!verify_inclusion(bad_leaf, &[], &[], &root));
    }

    /// Helper: find the index of `pubkey` in the sorted-by-pubkey ordering of `leaves`.
    fn sorted_index_of(leaves: &[([u8; 32], u64)], pubkey: &[u8; 32]) -> usize {
        let mut sorted: Vec<[u8; 32]> = leaves.iter().map(|(p, _)| *p).collect();
        sorted.sort();
        sorted.iter().position(|p| p == pubkey).unwrap()
    }
}
